File: iq/29/rotated/find.go
// Copyright 2025 Samvel Khalatyan. All rights reserved.

package rotated

type Index int

const IndexError = -1

func Find(nn []int, x int) Index {
	left := 0
	right := len(nn)
	isRotated := func() bool {
		return nn[left] > nn[right-1]
	}
	canMoveRight := func(x int) bool {
		if !isRotated() {
			return false
		}
		return x <= nn[right-1]
	}
	moveRight := func(mid int) {
		left = mid + 1
	}
	canMoveLeft := func(x int) bool {
		if !isRotated() {
			return false
		}
		return nn[left] <= x
	}
	moveLeft := func(mid int) {
		right = mid
	}
	for left < right {
		mid := left + (right-left)/2
		n := nn[mid]
		switch {
		case x < n:
			if canMoveRight(x) {
				moveRight(mid)
				break
			}
			moveLeft(mid)
		case x > n:
			if canMoveLeft(x) {
				moveLeft(mid)
				right = mid
				break
			}
			moveRight(mid)
		default:
			return Index(mid)
		}
	}
	return IndexError
}

File: iq/29/rotated/find.cc
// Copyright 2025 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//
// clang-format off-next-line
//go:build ignore

#include "iq/29/rotated/find.h"

#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"

namespace iq::rotated {

absl::StatusOr<int> Find(const std::vector<int>& nn, int k) {
  std::size_t low = 0;
  std::size_t hi = nn.size();
  while (low < hi) {
    const std::size_t idx = low + (hi - low) / 2;
    if (int n = nn[idx]; k < n) {
      // May need to move right due to rotation
      if (k >= nn[low]) {
        hi = idx;
      } else {
        low = idx + 1;
      }
    } else if (k > n) {
      // May need to move left due to rotation
      if (k <= nn[hi - 1]) {
        low = idx + 1;
      } else {
        hi = idx;
      }
    } else {
      return idx;
    }
  }
  return absl::OutOfRangeError("out of range");
}

}  // namespace iq::rotated
